src/clj/rems/workflow/dynamic.clj

```diff
@@ -390,6 +435,48 @@
                              (assoc application :state ::approved)
                              injections))))))
 
+(deftest test-comment
+  (let [application {:state ::submitted
+                     :applicantuserid "applicant"
+                     :commenters #{}
+                     :workflow {:type :workflow/dynamic
+                                :handlers ["assistant"]}}
+        injections {:valid-user? #{"commenter" "commenter2"}}]
+    (testing "required :valid-user? injection"
+      (is (= {:errors [[:missing-injection :valid-user?]]}
+             (handle-command {:actor "assistant" :commenter "commenter" :type ::request-comment}
+                             application
+                             {}))))
+    (testing "commenter must be a valid user"
+      (is (= {:errors [[:invalid-user "commenter3"]]}
+             (handle-command {:actor "assistant" :commenter "commenter3" :type ::request-comment}
+                             application
+                             injections))))
+    (testing "commenting before ::request-comment should fail"
+      (is (= {:errors [:unauthorized]}
+             (handle-command {:actor "commenter" :decision :approved :type ::comment}
+                             application
+                             injections))))
+    (let [requested (apply-commands application
+                                    [{:actor "assistant" :commenter "commenter" :type ::request-comment}
+                                     {:actor "assistant" :commenter "commenter2" :type ::request-comment}]
+                                    injections)]
+      (testing "request comment succesfully"
+        (is (= #{"commenter2" "commenter"} (:commenters requested))))
+      (testing "only the requested commenter can comment"
+        (is (= {:errors [:unauthorized]}
+               (handle-command {:actor "commenter3" :comment "..." :type ::comment}
```

Comment:
"commenter3" is not a valid user, so this command has multiple reasons to fail. Please add "commenter3" to the whitelist in injections (and use e.g. "invaliduser" in the test for valid user).